### PR TITLE
metrics: refactor metrics (part II) 

### PR DESCRIPTION
**Pull request by holiman**
This is a follow-up to https://github.com/ethereum/go-ethereum/pull/28031

### Split up interfaces, write vs read

The changes look large, but they're not actually _that_ invasive. The interfaces have been split up into one write-interface and one read-interface, with `Snapshot` being the gateway from write to read. This simplifies the semantics _a lot_. 

Example of splitting up an interface into one readonly 'snapshot' part, and one updatable writeonly part: 

```golang
type MeterSnapshot interface {
	Count() int64
	Rate1() float64
	Rate5() float64
	Rate15() float64
	RateMean() float64
}

// Meters count events to produce exponentially-weighted moving average rates
// at one-, five-, and fifteen-minutes and a mean rate.
type Meter interface {
	Mark(int64)
	Snapshot() MeterSnapshot
	Stop()
}
```

### A note about concurrency

This PR makes the concurrency model clearer. We have actual meters and snapshot of meters. The `meter` is the thing which can be accessed from the registry, and updates can be made to it. 

- For all `meters`, (`Gauge`, `Timer` etc), it is assumed that they are accessed by different threads, making updates. Therefore, all `meters` update-methods (`Inc`, `Add`, `Update`, `Clear` etc) need to be concurrency-safe. 
- All `meters` have a `Snapshot()` method. This method is _usually_ called from one thread, a backend-exporter. But it's fully possible to have several exporters simultaneously: therefore this method should also be concurrency-safe. 

TLDR: `meter`s are accessible via registry, all their methods must be concurrency-safe. 

For all `Snapshot`s, it is assumed that an individual exporter-thread has obtained a `meter` from the registry, and called the `Snapshot` method to obtain a readonly snapshot. This snapshot is _not_ guaranteed to be concurrency-safe. There's no need for a snapshot to be concurrency-safe, since exporters should not share snapshots. 

Note, though: that by happenstance a lot of the snapshots _are_ concurrency-safe, being unmutable minimal representations of a value. Only the more complex ones are _not_ threadsafe, those that lazily calculate things like `Variance()`, `Mean()`.

Example of how a background exporter typically works, obtaining the snapshot and sequentially accessing the non-threadsafe methods in it: 
```golang
		ms := metric.Snapshot()
                ...
		fields := map[string]interface{}{
			"count":    ms.Count(),
			"max":      ms.Max(),
			"mean":     ms.Mean(),
			"min":      ms.Min(),
			"stddev":   ms.StdDev(),
			"variance": ms.Variance(),
```

TLDR: `snapshots` are not guaranteed to be concurrency-safe (but often are).

### Sample changes

I also changed the `Sample` type: previously, it iterated the samples fully every time `Mean()`,`Sum()`, `Min()` or `Max()` was invoked. Since we now have readonly base data, we can just iterate it once, in the constructor, and set all four values at once. 

The same thing has been done for runtimehistogram. 

### ResettingTimer API

Back when ResettingTImer was implemented, as part of https://github.com/ethereum/go-ethereum/pull/15910, Anton implemented a `Percentiles` on the new type. However, the method did not conform to the other existing types which also had a `Percentiles`. 

1. The existing ones, on input, took `0.5` to mean `50%`. Anton used `50` to mean `50%`. 
2. The existing ones returned `float64` outputs, thus interpolating between values. A value-set of `0, 10`, at `50%` would return `5`, whereas Anton's would return either `0` or `10`. 

This PR removes the 'new' version, and uses only the 'legacy' percentiles, also for the ResettingTimer type. 

The resetting timer snapshot was also defined so that it would expose the internal values. This has been removed, and getters for `Max, Min, Mean` have been added instead. 

### Unexport types

A lot of types were exported, but do not need to be. This PR unexports quite a lot of them.

### CODE REVIEW

**Comment by holiman** [comment]
This is now running on `05/06`
```
Sep 01 19:45:02 bench05.ethdevops.io geth INFO [09-01|17:45:02.776] Starting peer-to-peer node instance=Geth/v1.13.0-unstable-40f291b0-20230901/linux-amd64/go1.21.0
Sep 01 19:45:11 bench06.ethdevops.io geth INFO [09-01|17:45:11.310] Starting peer-to-peer node instance=Geth/v1.13.0-unstable-28857080-20230831/linux-amd64/go1.21.0
```
So this PR on `bench05`, master on `bench06`. 
I had to revert the type-change resetting timer, not because it was wrong, but because the earlier code has used the `i`-format (non-default integer format), making it not possible to switch to using the default (float) format: 

```
Sep 01 19:43:18 bench05.ethdevops.io geth WARN [09-01|17:43:18.263] Unable to send to InfluxDB err="{\"error\":\"partial write: field type conflict: input field \\\"p50\\\" on measurement \\\"geth.hashdb/memcache/gc/time.span\\\" is type float, already exists as type integer dropped=2\"}\n"
```
This does look a bit odd; here's a 12-hour chart, `master`:
Screenshot 2023-09-02 at 12-54-42 Single Geth - Grafana
vs this PR: 
Screenshot 2023-09-02 at 12-55-32 Single Geth - Grafana
The latter is very empty. Needs to be investigated.

**Comment by holiman** [comment]
I made a little custom binary which sent in fake traffic metrics. The `master` reported `GetBlockHeadersMsg` stats, the "this PR" code reported `GetBlockBodiesMsg` stats. Then I ran them at the same time, reporting against the same host. 

The faked metrics were more or less similar distribution, but there's randomness so they aren't identical. 

Screenshot 2023-09-03 at 20-10-48 Single Geth - Grafana

So left is `master`, right is this PR. 

As far as I can tell, this PR does not modify the behaviour of the traffic tracker histogram charts.

**Comment by holiman** [comment]
Some time later 

Screenshot 2023-09-03 at 20-41-36 Single Geth - Grafana

**Comment by holiman** [comment]
> This does look a bit odd; here's a 12-hour chart, master:

I think I know the reason for this now: those charts only take into account traffic on `eth/66`, not `67` or `68`. So if the bulk of the peers are `67/68` the charts won't have a lot of content. So it varies from time to time

(@karalabe how about changing it so that all meters like `geth.p2p/wait/eth/66/0x05.histogram` becomes `geth.p2p/wait/eth/0x05.histogram`, disregarding `version`?

File: metrics/ewma.go [hunk 1]
```diff
@@ -36,78 +39,70 @@ func NewEWMA15() EWMA {
 	return NewEWMA(1 - math.Exp(-5.0/60.0/15))
 }
 
-// EWMASnapshot is a read-only copy of another EWMA.
-type EWMASnapshot float64
+// eWMASnapshot is a read-only copy of another EWMA.
+type eWMASnapshot float64
```
Comment (lightclient) [inline metrics/ewma.go:43]:
```suggestion
type ewmaSnapshot float64
```
You may consider this form instead. It seems less common to break an acronym up by case, e.g. see `HTTP` - it doesn't often appear `hTTP`.

File: metrics/gauge_float64.go [hunk 2]
```diff
@@ -39,43 +42,17 @@ func NewRegisteredGaugeFloat64(name string, r Registry) GaugeFloat64 {
...
-}
+// gaugeFloat64Snapshot is a read-only copy of another GaugeFloat64.
+type gaugeFloat64Snapshot float64
 
 // Value returns the value at the time the snapshot was taken.
-func (g GaugeFloat64Snapshot) Value() float64 { return float64(g) }
+func (g gaugeFloat64Snapshot) Value() float64 { return float64(g) }
 
 // NilGauge is a no-op Gauge.
 type NilGaugeFloat64 struct{}
```
Comment (lightclient) [inline metrics/gauge_float64.go:52]: Should we roll this gauge up by remove the extra spacing + delete the method comments?

File: metrics/influxdb/testdata/influxdbv1.want [hunk 3]
```diff
@@ -1,9 +1,11 @@
...
 goth.test/counter.count value=12345 978307200000000000
 goth.test/counter_float64.count value=54321.98 978307200000000000
 goth.test/gauge.gauge value=23456i 978307200000000000
 goth.test/gauge_float64.gauge value=34567.89 978307200000000000
 goth.test/gauge_info.gauge value="{\"arch\":\"amd64\",\"commit\":\"7caa2d8163ae3132c1c2d6978c76610caee2d949\",\"os\":\"linux\",\"protocol_versions\":\"64 65 66\",\"version\":\"1.10.18-unstable\"}" 978307200000000000
 goth.test/histogram.histogram count=3i,max=3i,mean=2,min=1i,p25=1,p50=2,p75=3,p95=3,p99=3,p999=3,p9999=3,stddev=0.816496580927726,variance=0.6666666666666666 978307200000000000
 goth.test/meter.meter count=0i,m1=0,m15=0,m5=0,mean=0 978307200000000000
-goth.test/resetting_timer.span count=6i,max=120000000i,mean=30000000,min=10000000i,p50=12000000i,p95=120000000i,p99=120000000i 978307200000000000
+goth.test/resetting_timer.span count=6i,max=120000000i,mean=30000000,min=10000000i,p50=12500000i,p95=120000000i,p99=120000000i 978307200000000000
```
Comment (lightclient) [inline metrics/influxdb/testdata/influxdbv1.want:10]: Why the update here?

**Review by lightclient — COMMENTED** [review]
Overall looks really nice. Great simplification of the interfaces!

File: metrics/sample.go [hunk 4]
```diff
@@ -77,96 +79,36 @@ func (s *ExpDecaySample) Clear() {
...
-	return s.values.Size()
-}
-
 // Snapshot returns a read-only copy of the sample.
-func (s *ExpDecaySample) Snapshot() Sample {
+func (s *ExpDecaySample) Snapshot() SampleSnapshot {
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 	vals := s.values.Values()
 	values := make([]int64, len(vals))
```
[the remark this answers is not in the document]
Comment (holiman) [reply metrics/sample.go:87]: Well, they're not of the same type, so it's not possible to mix them up...

File: metrics/influxdb/testdata/influxdbv1.want (hunk 3 above)
Comment (holiman) [reply metrics/influxdb/testdata/influxdbv1.want:10]: Good question! 
This is the resetting timer, 

> The existing ones returned float64 outputs, thus interpolating between values. A value-set of 0, 10, at 50% would return 5, whereas Anton's would return either 0 or 10.

So these new values are interpolated

File: metrics/gauge.go [hunk 5]
```diff
@@ -105,18 +65,20 @@ type StandardGauge struct {
...
 	g.value.Store(v)
 }
 
-// Value returns the gauge's current value.
-func (g *StandardGauge) Value() int64 {
-	return g.value.Load()
+// Update updates the gauge's value if v is larger then the current valie.
+func (g *StandardGauge) UpdateIfGt(v int64) {
+	if g.value.Load() < v {
+		g.value.Store(v)
```
[the remark this answers is not in the document]
Comment (holiman) [reply metrics/gauge.go:80]: Yes, this is a little bit racy (semantically racy, not crashy-racy). In practice, this is only used for deleting storage during destruction, and it's not multithreaded. 
A "proper" implementation should do something like this, I guess:  
```golang
	for{
		exist := g.value.Load()
		if exist>= v {
			break
		}
		if g.value.CompareAndSwap(exist, v){
			break
		}
	}
```

File: metrics/sample_test.go [hunk 6]
```diff
@@ -79,65 +87,42 @@ func BenchmarkUniformSample1028(b *testing.B) {
...
-		t.Errorf("s.Size(): 10 != %v\n", size)
-	}
-	if l := len(s.Values()); l != 10 {
-		t.Errorf("len(s.Values()): 10 != %v\n", l)
-	}
-	for _, v := range s.Values() {
-		if v > 10 || v < 0 {
-			t.Errorf("out of range [0, 10): %v\n", v)
-		}
+func min(a, b int) int {
```
[the remark this answers is not in the document]
Comment (holiman) [reply metrics/sample_test.go:90]: yup

File: metrics/gauge.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (holiman) [reply metrics/gauge.go]: Added this now

File: metrics/sample.go [hunk 7]
```diff
@@ -202,207 +144,155 @@ func (s *ExpDecaySample) update(t time.Time, v int64) {
...
+	sum  int64
 }
 
-func NewSampleSnapshot(count int64, values []int64) *SampleSnapshot {
-	return &SampleSnapshot{
+// newSampleSnapshotPrecalculated creates a read-only sampleSnapShot, using
+// precalculated sums to avoid iterating the values
+func newSampleSnapshotPrecalculated(count int64, values []int64,
+	min, max, sum int64) *sampleSnapshot {
+	if len(values) == 0 {
```
[the remark this answers is not in the document]
Comment (lightclient) [reply metrics/sample.go:200]: I don't really ever break function signatures, but when I do I usually to this style:

```
func newSampleSnapshotPrecalculated(
        count int64,
        values []int64,
	min, 
        max,
        sum int64
) *sampleSnapshot {
```

File: metrics/testdata/opentsb.want [hunk 8]
```diff
@@ -1,4 +1,4 @@
-put pre.elite.count 978307200 0 host=hal9000
+put pre.elite.count 978307200 1337 host=hal9000
```
[the remark this answers is not in the document]
Comment (holiman) [reply metrics/testdata/opentsb.want:1]: This is a 
```
	NewRegisteredMeter("elite", r).Mark(1337)
```
Previously, the `StandardMeter` would not always count the things that were 'Mark':ed, it collected them int 'uncounted', until someone called `tick` after 5 seconds had elapsed. It was a bit wonky. With the writeable/readonly separation, a snapshot taken after a `Mark` does indeed get the correct value.

File: metrics/sample.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (lightclient) [reply metrics/sample.go:200]: really just a style pref i think though, what's here is fine

File: metrics/runtimehistogram.go [hunk 9]
```diff
@@ -35,130 +41,112 @@ func (h *runtimeHistogram) update(mh *metrics.Float64Histogram) {
...
+		count += int64(c)
+		midpoint := h.midpoint(i)
+		sum += midpoint * float64(c)
+		if c == 0 {
+			continue
+		}
 
-func (h *runtimeHistogramSnapshot) Snapshot() Histogram {
-	return h
+		if c > 0 && !minSet {
```
Comment (MariusVanDerWijden) [inline metrics/runtimehistogram.go:106]: isn't c always > 0 at this point?

File: metrics/sample.go [hunk 10]
```diff
@@ -202,207 +144,155 @@ func (s *ExpDecaySample) update(t time.Time, v int64) {
...
-func (s *SampleSnapshot) Values() []int64 {
+func (s *sampleSnapshot) Values() []int64 {
 	values := make([]int64, len(s.values))
 	copy(values, s.values)
 	return values
 }
 
 // Variance returns the variance of values at the time the snapshot was taken.
-func (s *SampleSnapshot) Variance() float64 { return SampleVariance(s.values) }
+func (s *sampleSnapshot) Variance() float64 { return SampleVariance(s.mean, s.values) }
```
[the remark this answers is not in the document]
Comment (holiman) [reply metrics/sample.go:281]: Good question, and I was about to say that there was no real point. Typically, an exporter does something like this: 

```
		ps := ms.Percentiles([]float64{0.25, 0.5, 0.75, 0.95, 0.99, 0.999, 0.9999})
		measurement := fmt.Sprintf("%s%s.histogram", namespace, name)
		fields := map[string]interface{}{
			"count":    ms.Count(),
			"max":      ms.Max(),
			"mean":     ms.Mean(),
			"min":      ms.Min(),
			"stddev":   ms.StdDev(),
			"variance": ms.Variance(),
			"p25":      ps[0],
			"p50":      ps[1],
			"p75":      ps[2],
			"p95":      ps[3],
			"p99":      ps[4],
			"p999":     ps[5],
			"p9999":    ps[6],
		}
```
That is, invoke all the methods sequentially. In that case, we don't want to iterate N times, preferrably only once. Although, in the case of variance, we need to do it twice, first to obtain the mean. What is _not_ expected is that an exporter would invoke `Variance()` _twice_. However, `StdDev` and `Variance` are basically the same, and caching _would_ be beneficial. 

Good find!

File: metrics/runtimehistogram.go (hunk 9 above)
Comment (holiman) [reply metrics/runtimehistogram.go:106]: Yes

Comment (holiman) [reply metrics/runtimehistogram.go]: Thanks!

File: metrics/sample.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (lightclient) [reply metrics/sample.go:87]: It just doesn't read so nicely. Not too important, but a couple alternatives:

```go
values := make([]int64, len(s.values.Values()))
var ( ... )
for i, item := range s.values.Values() {
...
```

or

```
var ( 
    values []int64
    ...
)
for i, item := range s.values.Values() {
    v := item.v
    values = append(values, v)
    ...
```
But if you don't feel it's worth changing, feel free to mark this as resolved.

File: metrics/sample.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (holiman) [reply metrics/sample.go:201]: That's not valid go-code. This is, but it's more or less the same? 
```golang
func newSampleSnapshotPrecalculated(
	count int64,
	values []int64,
	min,
	max,
	sum int64) *sampleSnapshot {
	if len(values) == 0 {
```

[the remark this answers is not in the document]
Comment (holiman) [reply metrics/sample.go:200]: Ok, removed the break
